<p align = "center" draggable=”false” ><img src="https://user-images.githubusercontent.com/37101144/161836199-fdb0219d-0361-4988-bf26-48b0fad160a3.png" 
     width="200px"
     height="auto"/>
</p>

# <h1 align="center" id="heading">Sentiment Analysis of Reddit Data using Reddit API</h1>

In this live coding session, we leverage the Python Reddit API Wrapper (`PRAW`) to retrieve data from subreddits on [Reddit](https://www.reddit.com), and perform sentiment analysis using [`pipelines`](https://huggingface.co/docs/transformers/main_classes/pipelines) from [HuggingFace ( 🤗 the GitHub of Machine Learning )](https://techcrunch.com/2022/05/09/hugging-face-reaches-2-billion-valuation-to-build-the-github-of-machine-learning/), powered by [transformer](https://arxiv.org/pdf/1706.03762.pdf).

## Objectives

At the end of the session, you will 

- know how to work with APIs
- feel more comfortable navigating thru documentation, even inspecting the source code
- understand what a `pipeline` object is in HuggingFace
- perform sentiment analysis using `pipeline`
- run a python script in command line and get the results

## How to Submit

- At the end of each task, commit* the work into the repository you created before the assignment
- After completing all three tasks, make sure to push the notebook containing all code blocks and output cells to your repository you created before the assignment
- Submit the link to the notebook in Canvas

\***NEVER** commit a notebook displaying errors unless it is instructed otherwise. However, commit often; recall git ABC = **A**lways **B**e **C**ommitting.

## Tasks

### Task I: Instantiate a Reddit API Object

The first task is to instantiate a Reddit API object using [PRAW](https://praw.readthedocs.io/en/stable/), through which you will retrieve data. PRAW is a wrapper for [Reddit API](https://www.reddit.com/dev/api) that makes interacting with the Reddit API easier unless you are already an expert of [`requests`](https://docs.python-requests.org/en/latest/).

#### 1. Install packages

Please ensure you've ran all the cells in the `imports.ipynb`, located [here](https://github.com/FourthBrain/MLE-8/blob/main/assignments/week-3-analyze-sentiment-subreddit/imports.ipynb), to make sure you have all the required packages for today's assignment.

####  2. Create a new app on Reddit 

Create a new app on Reddit and save secret tokens; refer to [post in medium](https://towardsdatascience.com/how-to-use-the-reddit-api-in-python-5e05ddfd1e5c) for more details.

- Create a Reddit account if you don't have one, log into your account.
- To access the API, we need create an app. Slight updates, on the website, you need to navigate to `preference` > `app`, or click [this link](https://www.reddit.com/prefs/apps) and scroll all the way down. 
- Click to create a new app, fill in the **name**, choose `script`, fill in  **description** and **redirect uri** ( The redirect URI is where the user is sent after they've granted OAuth access to your application (more info [here](https://github.com/reddit-archive/reddit/wiki/OAuth2)) For our purpose, you can enter some random url, e.g., www.google.com; as shown below.


    <img src="https://miro.medium.com/max/700/1*lRBvxpIe8J2nZYJ6ucMgHA.png" width="500"/>
- Jot down `client_id` (left upper corner) and `client_secret` 

    NOTE: CLIENT_ID refers to 'personal use script" and CLIENT_SECRET to secret.
    
    <div>
    <img src="https://miro.medium.com/max/700/1*7cGAKth1PMrEf2sHcQWPoA.png" width="300"/>
    </div>

- Create `secrets_reddit.py` in the same directory with this notebook, fill in `client_id` and `secret_id` obtained from the last step. We will need to import those constants in the next step.
    ```
    REDDIT_API_CLIENT_ID = "client_id"
    REDDIT_API_CLIENT_SECRET = "secret_id"
    REDDIT_API_USER_AGENT = "any string except bot; ex. My User Agent"
    ```
- Add `secrets_reddit.py` to your `.gitignore` file if not already done. NEVER push credentials to a repo, private or public. 

#### 3. Instantiate a `Reddit` object

Now you are ready to create a read-only `Reddit` instance. Refer to [documentation](https://praw.readthedocs.io/en/stable/code_overview/reddit_instance.html) when necessary.

In [1]:
import sys

sys.path.append('/opt/anaconda3/envs/reddit-sa/lib/python3.8/site-packages')
print(sys.path)

['/Users/mattrobinson/fourth-brain/MLE-10/assignments/week-03-reddit-sentiment-analysis/nb', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python38.zip', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/lib-dynload', '', '/Users/mattrobinson/Library/Python/3.8/lib/python/site-packages', '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/site-packages', '/opt/anaconda3/envs/reddit-sa/lib/python3.8/site-packages']


In [2]:
import praw
from praw.models import SubredditHelper
import reddit_secrets
import pandas as pd

# Create a Reddit object which allows us to interact with the Reddit API
reddit = praw.Reddit(
    client_id=reddit_secrets.REDDIT_API_CLIENT_ID,
    client_secret=reddit_secrets.REDDIT_API_CLIENT_SECRET,
    user_agent=reddit_secrets.REDDIT_API_USER_AGENT
)

In [3]:
print(reddit) 

<details>
<summary>Expected output:</summary>   

```<praw.reddit.Reddit object at 0x10f8a0ac0>```
</details>

#### 4. Instantiate a `subreddit` object

Lastly, create a `subreddit` object for your favorite subreddit and inspect the object. The expected output you will see ar from `r/machinelearning` unless otherwise specified.

In [4]:
subreddit_name = "machinelearning"
subreddit = reddit.subreddit(subreddit_name)
print(dir(subreddit))

['MESSAGE_PREFIX', 'STR_FIELD', 'VALID_TIME_FILTERS', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_convert_to_fancypants', '_create_or_update', '_fetch', '_fetch_data', '_fetch_info', '_fetched', '_kind', '_parse_xml_response', '_path', '_prepare', '_read_and_post_media', '_reddit', '_reset_attributes', '_safely_add_arguments', '_submission_class', '_submit_media', '_subreddit_collections_class', '_subreddit_list', '_upload_inline_media', '_upload_media', '_url_parts', '_validate_gallery', '_validate_inline_media', '_validate_time_filter', 'banned', 'collections', 'comments', 'contributor', 'controversial', 'display_name', 'emoji', 'filters', 'flair', 'fullname', 'gilded', 'h

What is the display name of the subreddit?

In [5]:
# YOUR CODE HERE

ml_subreddit = subreddit.display_name
print(ml_subreddit)
# >> machinelearning

machinelearning


<details>
<summary>Expected output:</summary>   

    machinelearning
</details>

How about its title, is it different from the display name?

In [6]:
# YOUR CODE HERE
subreddit.title

'Machine Learning'

Answer: Yes, the subreddit title attribute is formatted as a Proper Noun.

<details>
<summary>Expected output:</summary>   

    Machine Learning
</details>

Print out the description of the subreddit:

In [7]:
subreddit.description

'**[Rules For Posts](https://www.reddit.com/r/MachineLearning/about/rules/)**\n--------\n+[Research](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AResearch)\n--------\n+[Discussion](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3ADiscussion)\n--------\n+[Project](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AProject)\n--------\n+[News](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3ANews)\n--------\n***[@slashML on Twitter](https://twitter.com/slashML)***\n--------\n***[Chat with us on Slack](https://join.slack.com/t/rml-talk/shared_invite/enQtNjkyMzI3NjA2NTY2LWY0ZmRjZjNhYjI5NzYwM2Y0YzZhZWNiODQ3ZGFjYmI2NTU3YjE1ZDU5MzM2ZTQ4ZGJmOTFmNWVkMzFiMzVhYjg)***\n--------\n**Beginners:**\n--------\nPlease have a look at [our FAQ and Link-Collection](http://www.reddit.com/r/MachineLearning/wiki/index)\n\n[Metacademy](http://www.metacademy.org) is a great resourc

<details>
<summary>Expected output:</summary>

    **[Rules For Posts](https://www.reddit.com/r/MachineLearning/about/rules/)**
    --------
    +[Research](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AResearch)
    --------
    +[Discussion](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3ADiscussion)
    --------
    +[Project](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AProject)
    --------
    +[News](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict
</details>

### Task II: Parse comments

#### 1. Top Posts of All Time

Find titles of top 10 posts of **all time** from your favorite subreddit. Refer to [Obtain Submission Instances from a Subreddit Section](https://praw.readthedocs.io/en/stable/getting_started/quick_start.html)) if necessary. Verify if the titles match what you read on Reddit.

In [8]:
# try run this line, what do you see? press q once you are done
?subreddit.top 

In [9]:
# get top ten posts (it's a generator)
top_ten_gen = reddit.subreddit(f"{ml_subreddit}").top(time_filter="all",limit=10)

# init list for posts
posts = []

# get each post's title and author
for post in top_ten_gen:
    posts.append(
        (post.title,post.author)
    ) 
    
# dataframe for a more neat display
posts_df = pd.DataFrame(posts,index = [x for x in range(0,len(posts))], columns=["Post","Author"])
posts_df

Post            Author
0  [Project] From books to presentations in 10s w...       cyrildiagne
1  [D] A Demo from 1993 of 32-year-old Yann LeCun...      TheInsaneApp
2  [R] First Order Motion Model applied to animat...  programmerChilli
3  [N] AI can turn old photos into moving Images ...      TheInsaneApp
4  [D] This AI reveals how much time politicians ...      TheInsaneApp
5               [D] Types of Machine Learning Papers      TheInsaneApp
6  [D] The machine learning community has a toxic...      yusuf-bengio
7  I made a robot that punishes me if it detects ...          _ayushp_
8  [Project] NEW PYTHON PACKAGE: Sync GAN Art to ...          mencil47
9  [P] Using oil portraits and First Order Model ...       Enguzelharf

<details> <summary>Expected output:</summary>

    [Project] From books to presentations in 10s with AR + ML
    [D] A Demo from 1993 of 32-year-old Yann LeCun showing off the World's first Convolutional Network for Text Recognition
    [R] First Order Motion Model applied to animate paintings
    [N] AI can turn old photos into moving Images / Link is given in the comments - You can also turn your old photo like this
    [D] This AI reveals how much time politicians stare at their phone at work
    [D] Types of Machine Learning Papers
    [D] The machine learning community has a toxicity problem
    [Project] NEW PYTHON PACKAGE: Sync GAN Art to Music with "Lucid Sonic Dreams"! (Link in Comments)
    [P] Using oil portraits and First Order Model to bring the paintings back to life
    [D] Convolution Neural Network Visualization - Made with Unity 3D and lots of Code / source - stefsietz (IG)    
</details>

#### 2. Top 10 Posts of This Week

What are the titles of the top 10 posts of **this week** from your favorite subreddit?

In [10]:
# YOUR CODE HERE
# get top ten posts from this week (it's a generator)
top_ten_gen = reddit.subreddit(f"{ml_subreddit}").top(time_filter="week",limit=10)

# init list for posts
posts = []

# get each post's title and author
for post in top_ten_gen:
    posts.append(
        (post.title,post.author,post.score)
    ) 
    
# dataframe for a more neat display, including upvote score
posts_df = pd.DataFrame(posts,index = [x for x in range(0,len(posts))], columns=["Post","Author","Upvote Score"])
posts_df

Post                Author  \
0  [P] Finetuned Diffusion: multiple fine-tuned S...  Illustrious_Row_9971   
1  [D] DALL·E to be made available as API, OpenAI...          TiredOldCrow   
2  [P] Made a text generation model to extend sta...     Neat-Delivery4741   
3  [R] APPLE research: GAUDI — a neural architect...      SpatialComputing   
4  [P] Learn diffusion models with Hugging Face c...                lewtun   
5  [News] The Stack: 3 TB of permissively license...       Singularian2501   
6  [R] Reincarnating Reinforcement Learning (Neur...  smallest_meta_review   
7  [P] Transcribe any podcast episode in just 1 m...         thundergolfer   
8  [N] Adversarial Policies Beat Professional-Lev...                xutw21   
9  [P] Fine Tuning Stable Diffusion: Naruto Chara...            mippie_moe   

   Upvote Score  
0          1101  
1           412  
2           392  
3           378  
4           314  
5           301  
6           242  
7           200  
8           170  
9           164

<details><summary>Expected output:</summary>

    [N] Ian Goodfellow, Apple’s director of machine learning, is leaving the company due to its return to work policy. In a note to staff, he said “I believe strongly that more flexibility would have been the best policy for my team.” He was likely the company’s most cited ML expert.
    [R][P] Thin-Plate Spline Motion Model for Image Animation + Gradio Web Demo
    [P] I’ve been trying to understand the limits of some of the available machine learning models out there. Built an app that lets you try a mix of CLIP from Open AI + Apple’s version of MobileNet, and more directly on your phone's camera roll.
    [R] Meta is releasing a 175B parameter language model
    [N] Hugging Face raised $100M at $2B to double down on community, open-source & ethics
    [P] T-SNE to view and order your Spotify tracks
    [D] : HELP Finding a Book - A book written for Google Engineers about foundational Math to support ML
    [R] Scaled up CLIP-like model (~2B) shows 86% Zero-shot on Imagenet
    [D] Do you use NLTK or Spacy for text preprocessing?
    [D] Democratizing Diffusion Models - LDMs: High-Resolution Image Synthesis with Latent Diffusion Models, a 5-minute paper summary by Casual GAN Papers
</details>

💽❓ Data Question:

Check out what other attributes the `praw.models.Submission` class has in the [docs](https://praw.readthedocs.io/en/stable/code_overview/models/submission.html). 

1. After having a chance to look through the docs, is there any other information that you might want to extract? How might this additional data help you?
*Answer_1: It would be helpful to add some other attributes such as number of comments, an upvote score, and whether the post was marked as a spoiler. Perhaps looking at the posts that included the spoiler attribute vs those without it would give some indication as to the factors that would affect a post's popularity.*


Write a sample piece of code below extracting three additional pieces of information from the submission below.

In [11]:
# YOUR CODE HERE

def get_top_ten(subreddit_name):
    
    # get top ten posts from this week (it's a generator)
    top_ten_gen = reddit.subreddit(f"{subreddit_name}").top(time_filter="week",limit=10)

    # init list for posts
    posts = []

    # get each post's title and author
    for post in top_ten_gen:
        posts.append(
            (
                post.title,
                post.author,
                post.score,
                post.spoiler,
                post.num_comments
            )
        ) 

    # return dataframe for a more neat display, including upvote score
    posts_df = pd.DataFrame(posts,index = [x for x in range(0,len(posts))], columns=["Post","Author","Upvote Score","Spoiler","Num of Comments"])
    posts_df
    
    return posts_df

get_top_ten(ml_subreddit)

Post                Author  \
0  [P] Finetuned Diffusion: multiple fine-tuned S...  Illustrious_Row_9971   
1  [D] DALL·E to be made available as API, OpenAI...          TiredOldCrow   
2  [P] Made a text generation model to extend sta...     Neat-Delivery4741   
3  [R] APPLE research: GAUDI — a neural architect...      SpatialComputing   
4  [P] Learn diffusion models with Hugging Face c...                lewtun   
5  [News] The Stack: 3 TB of permissively license...       Singularian2501   
6  [R] Reincarnating Reinforcement Learning (Neur...  smallest_meta_review   
7  [P] Transcribe any podcast episode in just 1 m...         thundergolfer   
8  [N] Adversarial Policies Beat Professional-Lev...                xutw21   
9  [P] Fine Tuning Stable Diffusion: Naruto Chara...            mippie_moe   

   Upvote Score  Spoiler  Num of Comments  
0          1102    False               56  
1           412    False               59  
2           396    False               55  
3           380    False                7  
4           317    False               14  
5           299    False               32  
6           242    False               31  
7           197    False                9  
8           165    False               49  
9           160    False                8

💽❓ Data Question:

2. Is there any information available that might be a concern when it comes to Ethical Data?
*A2: Not really in my opinion. There is not any personal information, location information, or sensitive information available from this API from what I have seen so far. There might be an ethical question of using a reddit user's name, but I don't see any issues that could be questionable from an ethical perspective.*

#### 3. Comment Code

Add comments to the code block below to describe what each line of the code does (Refer to [Obtain Comment Instances Section](https://praw.readthedocs.io/en/stable/getting_started/quick_start.html) when necessary). The code is adapted from [this tutorial](https://praw.readthedocs.io/en/stable/tutorials/comments.html)

The purpose is 
1. to understand what the code is doing 
2. start to comment your code whenever it is not self-explantory if you have not (others will thank you, YOU will thank you later 😊) 

In [12]:
%%time
from praw.models import MoreComments

print(subreddit.title)

# store top comments in a list
top_comments = []

# iterate over each submission in an iterator of top 2 posts
for submission in subreddit.top(limit=2):
    print(submission)
    
    top_comments.append(submission)
    
#    for each top level comment
    for top_level_comment in submission.comments:

        # check if the top level comment has even more comments on it
        if isinstance(top_level_comment, MoreComments):
            continue
        
        # continue to the next comment...
        # save each comment to the top comment when
            top_comments.append(top_level_comment.body)
            
        
print(top_comments)

Machine Learning
gh1dj9
kuc6tz
[Submission(id='gh1dj9'), Submission(id='kuc6tz')]
CPU times: user 75 ms, sys: 11.9 ms, total: 86.9 ms
Wall time: 2.57 s


#### 4. Inspect Comments

How many comments did you extract from the last step? Examine a few comments. 

In [13]:
#YOUR CODE HERE  # the answer may vary 693 for r/machinelearning


# for each submission in top_comments
for submission in top_comments:
    
    # get the commentstree for each submision
    comments_list = submission.comments.list()
    
    
    # iterate through the list of comment tree lists
    for com_list in comments_list:
        
        top_comments.append(com_list)
        
        # examine the body of the text
        print(com_list.body)

        
    

Twitter thread: [https://twitter.com/cyrildiagne/status/1259441154606669824](https://twitter.com/cyrildiagne/status/1259441154606669824)

Code: [https://github.com/cyrildiagne/ar-cutpaste/tree/clipboard](https://github.com/cyrildiagne/ar-cutpaste/tree/clipboard)

Background removal is done with U^(2-Net) (Qin et Al, Pattern Recognition 2020): [https://github.com/NathanUA/U-2-Net](https://github.com/NathanUA/U-2-Net)

**/!\\ EDIT:** You can now subscribe to a beta program to get early access to the app: [https://arcopypaste.app](https://arcopypaste.app)  !
Simple yet very useful. Thank you for sharing the code.
The future 🤯
Ohh the nightmare of making this into a stable product... Enough to drive you mad just thinking about it
Almost guaranteed, Apple will copy your idea in 3, 2, 1....
Wtffff. Well that was incredible.
Apple can’t wait to steal this and not credit the creators
fantastic!
Why did the boxes in the diagram turn gray?
How does the Algorithm decide what it cuts out from the 

AttributeError: 'MoreComments' object has no attribute 'body'

In [ ]:
import random

random_comments = [random.choice(top_comments) for i in range(3)]

inspect_comments = []
for comment in random_comments:
    inspect_comments.append(comment.body)
    
inspect_comments

<details> <summary>Some of the comments from `r/machinelearning` subreddit are:</summary>

    ['Awesome visualisation',
    'Similar to a stack or connected neurons.',
    'Will this Turing pass the Turing Test?']
</details>

💽❓ Data Question:

3. After having a chance to review a few samples of 5 comments from the subreddit, what can you say about the data? 
*A3: I would say that this data is quite messy and can be a bit difficult to handle. Especially since Reddit is constantly evoling this API, even such that one of the core functions in this tutorial was no longer valid. However, after some trouble shooting I was able to get access to the data that I needed for this assignemnt.*

HINT: Think about the "cleanliness" of the data, the content of the data, think about what you're trying to do - how does this data line up with your goal?

#### 5. Extract Top Level Comment from Subreddit `TSLA`.

Write your code to extract top level comments from the top 10 topics of a time period, e.g., year, from subreddit `TSLA` and store them in a list `top_comments_tsla`.  

In [33]:
# YOUR CODE HERE

# modify top ten func from above
def mod_top_ten(subreddit_name):
    
    # get top ten posts from this week (it's a generator)
    top_ten_gen = reddit.subreddit(f"{subreddit_name}").top(time_filter="week",limit=10)

    # init list for posts
    posts = []

    # get each post's title and author
    for post in top_ten_gen:
        posts.append(
            post.selftext
        ) 
    
    return posts

top_comments_tsla = mod_top_ten("TSLA")

In [34]:
len(top_comments_tsla) # Expected: 174 for r/machinelearning

7

In [37]:
[random.choice(top_comments_tsla) for i in range(len(top_comments_tsla))]

['',
 '',
 '',
 '',
 " Hey everyone,\n\nI'm looking for feedback on a new platform we have been working on all year. It's called MoneyAdx and it's an investment community where people can share information and advice.\n\nHere is a link to our space for conversations related to [TSLA](https://www.moneyadx.com/stock/tsla). \n\nI'm still working on the site and I'm open to any suggestions or feedback you might have. So please check it out and let me know what you think.\n\nSpecifically, please \n\n* Let us know what features we are missing that other websites have. \n* Share your ideas on what would make this website more useful for you.\n* Or anything else you think we should know! \n\nYou can either post your feedback here or on the website itself. \n\nThanks in advance! \n\nLooking forward to hearing from you,\n\nRichard H. \n\nFounder | MoneyAdx \n\nWe strive to empower our readers with the knowledge and skills necessary to take control of their financial future.",
 "Tesla TSLA Could 

<details>
<summary>Some of the comments from `r/TSLA` subreddit:</summary>

    ['I bought puts',
    '100%',
    'Yes. And I’m bag holding 1200 calls for Friday and am close to throwing myself out the window']
</details>

💽❓ Data Question:

4. Now that you've had a chance to review another subreddits comments, do you see any differences in the kinds of comments either subreddit has - and how might this relate to bias?
*A4: There were some major quality issues with the posts of TSLA. A lot of them were adds and spam*

### Task III: Sentiment Analysis

Let us analyze the sentiment of comments scraped from `r/TSLA` using a pre-trained HuggingFace model to make the inference. Take a [Quick tour](https://huggingface.co/docs/transformers/quicktour). 

#### 1. Import `pipeline`

In [38]:
from transformers import pipeline # YOUR CODE HERE

#### 2. Create a Pipeline to Perform Task "sentiment-analysis"

In [39]:
sentiment_model = pipeline("sentiment-analysis") # YOUR CODE HERE

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


#### 3. Get one comment from list `top_comments_tsla` from Task II - 5.

In [48]:
import random
comment = random.choice(top_comments_tsla)

In [54]:
comment = 'Bury Burry!!!!!'

The example comment is: `'Bury Burry!!!!!'`. Print out what you get. For reproducibility, use the same comment in the next step; consider setting a seed.

#### 4. Make Inference!

In [57]:


sentiment = sentiment_model(comment) # YOUR CODE HERE 
sentiment

[{'label': 'NEGATIVE', 'score': 0.989326000213623}]

What is the type of the output `sentiment`?

```
YOUR ANSWER HERE

The answer is an output of the label (positive or negative) and the score (0 to ~1)
```

In [58]:
print(f'The comment: {comment}')
print(f'Predicted Label is {sentiment[0]["label"]} and the score is {sentiment[0]["score"]:.3f}')

The comment: Bury Burry!!!!!
Predicted Label is NEGATIVE and the score is 0.989


For the example comment, the output is:

    The comment: Bury Burry!!!!!
    Predicted Label is NEGATIVE and the score is 0.989

🖥️❓ Model Question:

1. What does the score represent?
*A1: The score represents an accuracy score for the model. *

### Task IV: Put All Together

Let's pull all the piece together, create a simple script that does 

- get the subreddit
- get comments from the top posts for given subreddit
- run sentiment analysis 

#### Complete the Script

Once you complete the code, running the following block writes the code into a new Python script and saves it as `top_tlsa_comment_sentiment.py` under the same directory with the notebook. 

In [75]:
%%writefile top_tlsa_comment_sentiment.py

import reddit_secrets
import random

from typing import Dict, List

from praw import Reddit
from praw.models.reddit.subreddit import Subreddit
from praw.models import MoreComments

from transformers import pipeline


def get_subreddit(display_name:str) -> Subreddit:
    """Get subreddit object from display name

    Args:
        display_name (str): [description]

    Returns:
        Subreddit: [description]
    """
    reddit = Reddit(
        client_id=reddit_secrets.REDDIT_API_CLIENT_ID,        
        client_secret=reddit_secrets.REDDIT_API_CLIENT_SECRET,
        user_agent=reddit_secrets.REDDIT_API_USER_AGENT
        )
    
    subreddit = reddit.subreddit("TSLA")
    return subreddit

def get_comments(subreddit:Subreddit, limit:int=3) -> List[str]:
    """ Get comments from subreddit

    Args:
        subreddit (Subreddit): [description]
        limit (int, optional): [description]. Defaults to 3.

    Returns:
        List[str]: List of comments
    """
    top_comments = []
    for submission in subreddit.top(limit=limit):
        for top_level_comment in submission.comments:
            if isinstance(top_level_comment, MoreComments):
                continue
            top_comments.append(top_level_comment.body)
    return top_comments

def run_sentiment_analysis(comment:str) -> Dict:
    """Run sentiment analysis on comment using default distilbert model
    
    Args:
        comment (str): [description]
        
    Returns:
        str: Sentiment analysis result
    """
    sentiment_model = pipeline("sentiment-analysis")
    sentiment = sentiment_model(comment)
    return sentiment[0]


if __name__ == '__main__':

    subreddit_display_name = "TSLA"
    subreddit = get_subreddit(subreddit_display_name)
    comments = get_comments(subreddit)
    comment = [comments[0]]
    sentiment = run_sentiment_analysis(comment)
    
    print(f'The comment: {comment}')
    print(f'Predicted Label is {sentiment["label"]} and the score is {sentiment["score"]:.3f}')

Overwriting top_tlsa_comment_sentiment.py


Run the following block to see the output.

In [76]:
!python top_tlsa_comment_sentiment.py

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
The comment: ['ho lee fuk \n\nyou got anymore insider information? 👀👀']
Predicted Label is NEGATIVE and the score is 0.994


<details><summary> Expected output:</summary>

    No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english)
    The comment: When is DOGE flying
    Predicted Label is POSITIVE and the score is 0.689
</details>

💽❓ Data Question:

5. Is the subreddit active? About how many posts or threads per day? How could you find this information?
*This subreddit is not too active anymore. The TSLA subreddit is basically all adds*

💽❓ Data Question:

6. Does there seem to be a large distribution of posters or a smaller concentration of posters who are very active? What kind of impact might this have on the data?

*The posts seemed to be heavily created by bots and this would disrupt any real analysis of the subreddit.*